src/core/tests/frontend/paddle/test_models/gen_scripts/generate_box_coder.py

```diff
@@ -0,0 +1,73 @@
+import numpy as np
+from save_model import saveModel
+import sys
+
+
+def box_coder(name: str, prior_box, prior_box_var, target_box, code_type, box_normalized, axis, dtype):
+    import paddle
+    import paddle.fluid as fluid
+    paddle.enable_static()
+    is_tensor = not isinstance(prior_box_var, list)
+    with paddle.static.program_guard(paddle.static.Program(), paddle.static.Program()):
+
+        prior_box_decode = paddle.static.data(name='prior_box',
+                                              shape=prior_box.shape,
+                                              dtype=dtype)
+        target_box_decode = paddle.static.data(name='target_box',
+                                               shape=target_box.shape,
+                                               dtype=dtype)
+        if is_tensor:
+            prior_box_var_decode = paddle.static.data(name='prior_box_var',
+                                                      shape=prior_box_var.shape,
+                                                      dtype=dtype)
+        else:
+            prior_box_var_decode = prior_box_var
+
+        out = fluid.layers.box_coder(prior_box=prior_box_decode,
+                                     prior_box_var=prior_box_var_decode,
+                                     target_box=target_box_decode,
+                                     code_type=code_type,
+                                     box_normalized=box_normalized,
+                                     axis=axis)
+
+        cpu = paddle.static.cpu_places(1)
+        exe = paddle.static.Executor(cpu[0])
+        # startup program will call initializer to initialize the parameters.
+        exe.run(paddle.static.default_startup_program())
+
+        feed_dict = {'prior_box': prior_box, 'target_box': target_box}
+        if is_tensor:
+            feed_dict['prior_box_var'] = prior_box_var
+
+        outs = exe.run(
+            feed=feed_dict,
+            fetch_list=[out])
+
+        saveModel(name, exe, feedkeys=[*feed_dict.keys()], fetchlist=[out], inputs=[*feed_dict.values()], outputs=[outs[0]],
```

Comment:
Please use exportModel() instead of saveModel() to cover the dynamic input shape cases. In such case of using a decomposition graph to map a paddle operator, it is prone to fail.

You could refer to https://github.com/openvinotoolkit/openvino/blob/141e1f8dd78cc2d22e512b29b0fbd2ef77537f15/src/core/tests/frontend/paddle/test_models/gen_scripts/generate_slice.py#L39 on how to use exportModel.

And you may refer to https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/test_box_coder_op.py on paddle unit tests.

Reply:
Hi, It seems that `exportModel` is used to cover dynamic input tensor case instead of dynamic input shape case?

Reply:
It is the same thing. Only once the model has dynamic input shape, it could accept dynamic input tensor as its input.

Anyway, please fix other requests first and leave the dynamism request last. We can see this dynamism support as a better-to-have feature.
It is a plus. @OpenVINO-dev-contest .

Reply:
It seems that paddle.fluid.layers.box_coder does not support dynamic graph.
I can't find any introduction about box_coder since paddle 1.8, it's when paddle only support static graph.

Reply:
Please refer to https://github.com/PaddlePaddle/PaddleDetection/blob/38f695ebdc09cdfeebc2870105442908b42f062a/ppdet/modeling/ops.py#L886
There are some ops in the unit tests like multiclass_nms and generateproposals uses PaddleDetection to create test cases.

Reply:
> Please refer to https://github.com/PaddlePaddle/PaddleDetection/blob/38f695ebdc09cdfeebc2870105442908b42f062a/ppdet/modeling/ops.py#L886 There are some ops in the unit tests like multiclass_nms and generateproposals uses PaddleDetection to create test cases.

Hi,
I'm wondering that what is the purpose of this.
I follow the implementation in PaddleDetection,  and use `exportModel` to convert dynamic function to static one.
```python
def ref_box_coder(prior_box,
                  prior_box_var,
                  target_box,
                  code_type="encode_center_size",
                  box_normalized=True,
                  axis=0,
                  name=None):

    check_variable_and_dtype(prior_box, 'prior_box', ['float32', 'float64'],
                             'box_coder')
    check_variable_and_dtype(target_box, 'target_box', ['float32', 'float64'],
                             'box_coder')

    if in_dygraph_mode():
        if isinstance(prior_box_var, Variable):
            output_box = fluid.core.ops.box_coder(
                prior_box, prior_box_var, target_box, "code_type", code_type,
                "box_normalized", box_normalized, "axis", axis)

        elif isinstance(prior_box_var, list):
            output_box = fluid.core.ops.box_coder(
                prior_box, None, target_box, "code_type", code_type,
                "box_normalized", box_normalized, "axis", axis, "variance",
                prior_box_var)
        else:
            raise TypeError(
                "Input variance of box_coder must be Variable or list")
        return output_box
    else:
        helper = LayerHelper("box_coder", **locals())

        output_box = helper.create_variable_for_type_inference(
            dtype=prior_box.dtype)

        inputs = {"PriorBox": prior_box, "TargetBox": target_box}
        attrs = {
            "code_type": code_type,
            "box_normalized": box_normalized,
            "axis": axis
        }
        if isinstance(prior_box_var, Variable):
            inputs['PriorBoxVar'] = prior_box_var
        elif isinstance(prior_box_var, list):
            attrs['variance'] = prior_box_var
        else:
            raise TypeError(
                "Input variance of box_coder must be Variable or list")
        helper.append_op(
            type="box_coder",
            inputs=inputs,
            attrs=attrs,
            outputs={"OutputBox": output_box})
        return output_box

def main():
    # For decode
    datatype = "float32"
    prior_box = np.random.random([8, 4]).astype(datatype)
    target_box = np.random.random([8, 4, 4]).astype(datatype)
    prior_box_var = [0.1, 0.1, 0.1, 0.1]
    code_type = "decode_center_size"
    box_normalized = True
    axis = 1
    paddle_out = box_coder("box_coder_1", prior_box, prior_box_var, target_box, code_type, box_normalized, axis,
                           datatype)

    axis = 0
    prior_box = np.random.random([4, 4]).astype(datatype)
    box_normalized = False
    prior_box_var = np.repeat(np.array([[0.1, 0.2, 0.1, 0.1]], dtype=np.float32), prior_box.shape[0], axis=0)
    paddle_out = box_coder("box_coder_2", prior_box, prior_box_var, target_box, code_type, box_normalized, axis,
                           datatype)

    box_normalized = True
    paddle_out = box_coder("box_coder_3", prior_box, prior_box_var, target_box, code_type, box_normalized, axis,
                           datatype)

def box_coder_dygraph():
    paddle.disable_static()

    @paddle.jit.to_static
    def test_model_1(prior_box, target_box):
        prior_box_var = [0.1, 0.1, 0.1, 0.1]
        code_type = "decode_center_size"
        box_normalized = True
        axis = 1
        out = ref_box_coder(prior_box, prior_box_var, target_box, code_type, box_normalized, axis)
        return out

    datatype = "float32"
    prior_box = paddle.rand(shape=[2, 4], dtype=datatype)
    target_box = paddle.rand(shape=[2, 4, 4], dtype=datatype)
    exportModel("box_coder_dygraph_1", test_model_1, [prior_box, target_box], target_dir=sys.argv[1])

    @paddle.jit.to_static
    def test_model_2(prior_box, target_box, prior_box_var):
        code_type = "decode_center_size"
        box_normalized = True
        axis = 0
        out = ref_box_coder(prior_box, prior_box_var, target_box, code_type, box_normalized, axis)
        return out

    datatype = "float32"
    prior_box = paddle.rand(shape=[4, 4], dtype=datatype)
    target_box = paddle.rand(shape=[8, 4, 4], dtype=datatype)
    prior_box_var = paddle.repeat_interleave(paddle.to_tensor([[0.1, 0.2, 0.1, 0.1]], dtype=datatype), prior_box.shape[0], axis=0)

    exportModel("box_coder_dygraph_2", test_model_2, [prior_box, target_box, prior_box_var], target_dir=sys.argv[1])

```
But `box_coder_dygraph_2` test failed, even though `box_coder_3` which has the same parameters settting and input shape can pass
![image](https://user-images.githubusercontent.com/72954905/187216918-7b4ba8f1-5c61-47ee-94b5-9d3580c71046.png)

Reply:
So you have succeeded to the paddle model generated, but one of them fails to run, right? What's the error you observed? You probably could add "--gtest_break_on_failure" to make the message visible when error happens.
Also, you can use netron or https://www.paddlepaddle.org.cn/paddle/visualdl/demo/graph to visualize the model.

Let me know if you need assistance on debugging.

btw, a slight reminder is, you don't have to implement "ref_box_coder". You could directly use it. Please refer to https://github.com/openvinotoolkit/openvino/blob/7f665502b622eb07b85168f400f3fafceb096142/src/core/tests/frontend/paddle/test_models/gen_scripts/generate_generate_proposal_v2.py#L22

Reply:
> So you have succeeded to the paddle model generated, but one of them fails to run, right?

Yes, the result: 
```bash
Note: Google Test filter = PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder*
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from PaddleFuzzyOpTest/FrontEndFuzzyOpTest
[ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_1
[       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_1 (398 ms)
[ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_2
[       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_2 (40 ms)
[ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_3
[       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_3 (25 ms)
[ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_1_box_coder_dygraph_1_pdmodel
[       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_1_box_coder_dygraph_1_pdmodel (23 ms)
[ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_2_box_coder_dygraph_2_pdmodel
0 is not close to 0.1000000014901161194 at index 0
0 is not close to 0.1000000014901161194 at index 1
0 is not close to 0.1000000014901161194 at index 2
0 is not close to 0.1000000014901161194 at index 3
0 is not close to 0.1000000014901161194 at index 16
0 is not close to 0.1000000014901161194 at index 17
0 is not close to 0.1000000014901161194 at index 18
0 is not close to 0.1000000014901161194 at index 19
0 is not close to 0.1000000014901161194 at index 32
0 is not close to 0.1000000014901161194 at index 33
0 is not close to 0.1000000014901161194 at index 34
0 is not close to 0.1000000014901161194 at index 35
0 is not close to 0.1000000014901161194 at index 48
0 is not close to 0.1000000014901161194 at index 49
0 is not close to 0.1000000014901161194 at index 50
0 is not close to 0.1000000014901161194 at index 51
0 is not close to 0.1000000014901161194 at index 64
0 is not close to 0.1000000014901161194 at index 65
0 is not close to 0.1000000014901161194 at index 66
0 is not close to 0.1000000014901161194 at index 67
0 is not close to 0.1000000014901161194 at index 80
0 is not close to 0.1000000014901161194 at index 81
0 is not close to 0.1000000014901161194 at index 82
0 is not close to 0.1000000014901161194 at index 83
0 is not close to 0.1000000014901161194 at index 96
0 is not close to 0.1000000014901161194 at index 97
0 is not close to 0.1000000014901161194 at index 98
0 is not close to 0.1000000014901161194 at index 99
0 is not close to 0.1000000014901161194 at index 112
0 is not close to 0.1000000014901161194 at index 113
0 is not close to 0.1000000014901161194 at index 114
0 is not close to 0.1000000014901161194 at index 115

/home/disk2/ray/workspace/StarFall.ov/src/tests/engines_util/../engines_util/test_case.hpp:212: Failure
Value of: res
  Actual: false (0 is not close to 0.1000000014901161194 at index 0
0 is not close to 0.1000000014901161194 at index 1
0 is not close to 0.1000000014901161194 at index 2
0 is not close to 0.1000000014901161194 at index 3
0 is not close to 0.1000000014901161194 at index 16
0 is not close to 0.1000000014901161194 at index 17
0 is not close to 0.1000000014901161194 at index 18
0 is not close to 0.1000000014901161194 at index 19
0 is not close to 0.1000000014901161194 at index 32
0 is not close to 0.1000000014901161194 at index 33
0 is not close to 0.1000000014901161194 at index 34
0 is not close to 0.1000000014901161194 at index 35
0 is not close to 0.1000000014901161194 at index 48
0 is not close to 0.1000000014901161194 at index 49
0 is not close to 0.1000000014901161194 at index 50
0 is not close to 0.1000000014901161194 at index 51
0 is not close to 0.1000000014901161194 at index 64
0 is not close to 0.1000000014901161194 at index 65
0 is not close to 0.1000000014901161194 at index 66
0 is not close to 0.1000000014901161194 at index 67
0 is not close to 0.1000000014901161194 at index 80
0 is not close to 0.1000000014901161194 at index 81
0 is not close to 0.1000000014901161194 at index 82
0 is not close to 0.1000000014901161194 at index 83
0 is not close to 0.1000000014901161194 at index 96
0 is not close to 0.1000000014901161194 at index 97
0 is not close to 0.1000000014901161194 at index 98
0 is not close to 0.1000000014901161194 at index 99
0 is not close to 0.1000000014901161194 at index 112
0 is not close to 0.1000000014901161194 at index 113
0 is not close to 0.1000000014901161194 at index 114
0 is not close to 0.1000000014901161194 at index 115
)
Expected: true
[  FAILED  ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_2_box_coder_dygraph_2_pdmodel, where GetParam() = ("paddle", "test_model_zoo/paddle_test_models/", "box_coder_dygraph_2/box_coder_dygraph_2.pdmodel") (26 ms)
[----------] 5 tests from PaddleFuzzyOpTest/FrontEndFuzzyOpTest (512 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (512 ms total)
[  PASSED  ] 4 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_2_box_coder_dygraph_2_pdmodel, where GetParam() = ("paddle", "test_model_zoo/paddle_test_models/", "box_coder_dygraph_2/box_coder_dygraph_2.pdmodel")
```

Reply:
> btw, a slight reminder is, you don't have to implement "ref_box_coder". You could directly use it. Please refer to
> 
> https://github.com/openvinotoolkit/openvino/blob/7f665502b622eb07b85168f400f3fafceb096142/src/core/tests/frontend/paddle/test_models/gen_scripts/generate_generate_proposal_v2.py#L22

down.

Reply:
Do you have new commits for it?

Reply:
> > So you have succeeded to the paddle model generated, but one of them fails to run, right?
> 
> Yes, the result:
> 
> ```shell
> Note: Google Test filter = PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder*
> [==========] Running 5 tests from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 5 tests from PaddleFuzzyOpTest/FrontEndFuzzyOpTest
> [ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_1
> [       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_1 (398 ms)
> [ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_2
> [       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_2 (40 ms)
> [ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_3
> [       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_3 (25 ms)
> [ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_1_box_coder_dygraph_1_pdmodel
> [       OK ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_1_box_coder_dygraph_1_pdmodel (23 ms)
> [ RUN      ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_2_box_coder_dygraph_2_pdmodel
> 0 is not close to 0.1000000014901161194 at index 0
> 0 is not close to 0.1000000014901161194 at index 1
> 0 is not close to 0.1000000014901161194 at index 2
> 0 is not close to 0.1000000014901161194 at index 3
> 0 is not close to 0.1000000014901161194 at index 16
> 0 is not close to 0.1000000014901161194 at index 17
> 0 is not close to 0.1000000014901161194 at index 18
> 0 is not close to 0.1000000014901161194 at index 19
> 0 is not close to 0.1000000014901161194 at index 32
> 0 is not close to 0.1000000014901161194 at index 33
> 0 is not close to 0.1000000014901161194 at index 34
> 0 is not close to 0.1000000014901161194 at index 35
> 0 is not close to 0.1000000014901161194 at index 48
> 0 is not close to 0.1000000014901161194 at index 49
> 0 is not close to 0.1000000014901161194 at index 50
> 0 is not close to 0.1000000014901161194 at index 51
> 0 is not close to 0.1000000014901161194 at index 64
> 0 is not close to 0.1000000014901161194 at index 65
> 0 is not close to 0.1000000014901161194 at index 66
> 0 is not close to 0.1000000014901161194 at index 67
> 0 is not close to 0.1000000014901161194 at index 80
> 0 is not close to 0.1000000014901161194 at index 81
> 0 is not close to 0.1000000014901161194 at index 82
> 0 is not close to 0.1000000014901161194 at index 83
> 0 is not close to 0.1000000014901161194 at index 96
> 0 is not close to 0.1000000014901161194 at index 97
> 0 is not close to 0.1000000014901161194 at index 98
> 0 is not close to 0.1000000014901161194 at index 99
> 0 is not close to 0.1000000014901161194 at index 112
> 0 is not close to 0.1000000014901161194 at index 113
> 0 is not close to 0.1000000014901161194 at index 114
> 0 is not close to 0.1000000014901161194 at index 115
> 
> /home/disk2/ray/workspace/StarFall.ov/src/tests/engines_util/../engines_util/test_case.hpp:212: Failure
> Value of: res
>   Actual: false (0 is not close to 0.1000000014901161194 at index 0
> 0 is not close to 0.1000000014901161194 at index 1
> 0 is not close to 0.1000000014901161194 at index 2
> 0 is not close to 0.1000000014901161194 at index 3
> 0 is not close to 0.1000000014901161194 at index 16
> 0 is not close to 0.1000000014901161194 at index 17
> 0 is not close to 0.1000000014901161194 at index 18
> 0 is not close to 0.1000000014901161194 at index 19
> 0 is not close to 0.1000000014901161194 at index 32
> 0 is not close to 0.1000000014901161194 at index 33
> 0 is not close to 0.1000000014901161194 at index 34
> 0 is not close to 0.1000000014901161194 at index 35
> 0 is not close to 0.1000000014901161194 at index 48
> 0 is not close to 0.1000000014901161194 at index 49
> 0 is not close to 0.1000000014901161194 at index 50
> 0 is not close to 0.1000000014901161194 at index 51
> 0 is not close to 0.1000000014901161194 at index 64
> 0 is not close to 0.1000000014901161194 at index 65
> 0 is not close to 0.1000000014901161194 at index 66
> 0 is not close to 0.1000000014901161194 at index 67
> 0 is not close to 0.1000000014901161194 at index 80
> 0 is not close to 0.1000000014901161194 at index 81
> 0 is not close to 0.1000000014901161194 at index 82
> 0 is not close to 0.1000000014901161194 at index 83
> 0 is not close to 0.1000000014901161194 at index 96
> 0 is not close to 0.1000000014901161194 at index 97
> 0 is not close to 0.1000000014901161194 at index 98
> 0 is not close to 0.1000000014901161194 at index 99
> 0 is not close to 0.1000000014901161194 at index 112
> 0 is not close to 0.1000000014901161194 at index 113
> 0 is not close to 0.1000000014901161194 at index 114
> 0 is not close to 0.1000000014901161194 at index 115
> )
> Expected: true
> [  FAILED  ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_2_box_coder_dygraph_2_pdmodel, where GetParam() = ("paddle", "test_model_zoo/paddle_test_models/", "box_coder_dygraph_2/box_coder_dygraph_2.pdmodel") (26 ms)
> [----------] 5 tests from PaddleFuzzyOpTest/FrontEndFuzzyOpTest (512 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 5 tests from 1 test suite ran. (512 ms total)
> [  PASSED  ] 4 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_box_coder_dygraph_2_box_coder_dygraph_2_pdmodel, where GetParam() = ("paddle", "test_model_zoo/paddle_test_models/", "box_coder_dygraph_2/box_coder_dygraph_2.pdmodel")
> ```

Do you meet problem to debug it? The problem might come from the op mapping itself or a potential bug in openvino. Will you help narrow down?

Reply:
> Do you meet problem to debug it? The problem might come from the op mapping itself or a potential bug in openvino. Will you help narrow down?

I wil try.